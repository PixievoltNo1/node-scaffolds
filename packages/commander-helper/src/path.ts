import path from 'path'

/**
 * Calc absolute path of p under the workspace
 *
 * @param workspace
 * @param targetPath
 */
export function absoluteOfWorkspace(
  workspace: string,
  targetPath?: string | null,
): string {
  if (targetPath == null) return workspace
  const absoluteDir: string = path.resolve(workspace, targetPath)
  return absoluteDir
}

/**
 * Calc relative path to workspace
 *
 * @param workspace
 * @param targetDir
 */
export function relativeOfWorkspace(
  workspace: string,
  targetDir: string,
): string {
  const absoluteDir = absoluteOfWorkspace(workspace, targetDir)
  const relativeDir: string = path.relative(workspace, absoluteDir)
  return path.normalize(relativeDir)
}
