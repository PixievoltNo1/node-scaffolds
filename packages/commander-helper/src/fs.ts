import { isFileSync } from '@guanghechen/file-helper'
import fs from 'fs-extra'
import yaml from 'js-yaml'
import path from 'path'

/**
 * Load configuration file with format .json / .yml / .yaml
 *
 * @param filepath
 * @param logger
 */
export async function loadJsonOrYaml(
  filepath: string,
  encoding = 'utf-8',
): Promise<unknown | never> {
  if (!isFileSync(filepath))
    throw new Error(`${filepath} is an invalid file path`)

  const loadContent = (): Promise<string> =>
    fs.readFile(filepath, encoding as BufferEncoding)

  let result: unknown
  const extname = path.extname(filepath)
  switch (extname) {
    case '.json':
      {
        const content: string = await loadContent()
        result = JSON.parse(content)
      }
      break
    case '.yml':
    case '.yaml':
      {
        const content: string = await loadContent()
        result = yaml.load(content, { filename: filepath, json: true })
      }
      break
    default:
      throw new Error(
        `Only files in .json / .yml / .ymal format are supported. filepath(${filepath}`,
      )
  }
  return result
}

/**
 * Load configuration file with format .json / .yml / .yaml  (synchronizing)
 *
 * @param filepath
 * @param logger
 */
export function loadJsonOrYamlSync(
  filepath: string,
  encoding = 'utf-8',
): unknown | never {
  if (!isFileSync(filepath))
    throw new Error(`${filepath} is an invalid file path`)

  const loadContent = (): string =>
    fs.readFileSync(filepath, encoding as BufferEncoding)

  let result: unknown
  const extname = path.extname(filepath)
  switch (extname) {
    case '.json':
      {
        const content: string = loadContent()
        result = JSON.parse(content)
      }
      break
    case '.yml':
    case '.yaml':
      {
        const content: string = loadContent()
        result = yaml.load(content, { filename: filepath, json: true })
      }
      break
    default:
      throw new Error(
        `Only files in .json / .yml / .ymal format are supported. filepath(${filepath}`,
      )
  }
  return result
}
