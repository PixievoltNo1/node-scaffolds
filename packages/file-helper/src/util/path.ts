import path from 'path'

/**
 * Calc absolute path of p under the workspace
 *
 * @param workspace
 * @param targetPath
 */
export function absoluteOfWorkspace(workspace: string, targetPath?: string | null): string {
  if (targetPath == null) return workspace
  const filepath: string = path.resolve(workspace, targetPath)
  return path.normalize(filepath)
}

/**
 * Calc relative path to workspace
 *
 * @param workspace
 * @param targetPath
 */
export function relativeOfWorkspace(workspace: string, targetPath: string): string {
  const absoluteDir = absoluteOfWorkspace(workspace, targetPath)
  const filepath: string = path.relative(workspace, absoluteDir)
  return path.normalize(filepath)
}
