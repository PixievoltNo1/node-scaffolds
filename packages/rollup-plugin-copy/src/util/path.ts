import path from 'path'

/**
 * Resolve the source path.
 * @param workspace
 * @param srcPath
 * @returns
 */
export const resolvePath = (workspace: string, srcPath: string): string => {
  const filepath = path.resolve(workspace, srcPath)
  return filepath
}

/**
 * Get the path relative to workspace.
 * @param workspace
 * @param filepath
 * @returns
 */
export const relativePath = (workspace: string, filepath: string): string => {
  const srcPath: string = filepath.startsWith(workspace)
    ? filepath.slice(workspace.length).replace(/^[/\\]/, '')
    : filepath
  return srcPath
}
