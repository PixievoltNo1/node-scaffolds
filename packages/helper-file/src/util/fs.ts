import fs from 'node:fs'
import path from 'node:path'

/**
 * Collect all files under the given directory.
 *
 * @param dir
 * @param predicate
 */
export async function collectAllFiles(
  dir: string,
  predicate?: ((p: string, stat: fs.Stats) => Promise<boolean> | boolean) | null,
): Promise<string[]> {
  const results: string[] = []
  await collect(dir)
  return results

  async function collect(filepath: string): Promise<void> {
    const stat = fs.statSync(filepath)
    if (stat.isDirectory()) {
      const filenames = fs.readdirSync(filepath)
      for (const filename of filenames) {
        const nextFilepath = path.join(filepath, filename)
        await collect(nextFilepath)
      }
      return
    }

    if (stat.isFile()) {
      if (predicate == null || (await predicate(filepath, stat))) {
        results.push(filepath)
      }
    }
  }
}

/**
 * Collect all files under the given directory.   (synchronizing)
 *
 * @param dir
 * @param predicate
 */
export function collectAllFilesSync(
  dir: string,
  predicate?: ((p: string, stat: fs.Stats) => boolean) | null,
): string[] {
  const results: string[] = []
  collect(dir)
  return results

  function collect(filepath: string): void {
    const stat = fs.statSync(filepath)
    if (stat.isDirectory()) {
      const filenames = fs.readdirSync(filepath)
      for (const filename of filenames) {
        const nextFilepath = path.join(filepath, filename)
        collect(nextFilepath)
      }
      return
    }
    if (stat.isFile()) {
      if (predicate == null || predicate(filepath, stat)) {
        results.push(filepath)
      }
    }
  }
}
