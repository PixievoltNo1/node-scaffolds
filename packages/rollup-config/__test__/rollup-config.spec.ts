import fs from 'fs-extra'
import path from 'path'
import { rollup } from 'rollup'
import type { OutputOptions, RollupOutput } from 'rollup'
import createRollupConfig from '../src'

// Resolve absolute dirPath of case
const resolveCaseDir = (title: string): string => path.resolve(__dirname, 'fixtures', title)

async function build(dependencies: string[]): Promise<RollupOutput[]> {
  // Create rollup config
  const config = createRollupConfig({
    manifest: {
      source: 'src/index.ts',
      main: 'lib/cjs/index.js',
      module: 'lib/esm/index.js',
      dependencies: dependencies,
    },
    pluginOptions: {
      typescriptOptions: { tsconfig: 'tsconfig.src.json' },
    },
  })
  const bundle = await rollup(config)

  const results: any[] = []
  for (const outputOptions of config.output as OutputOptions[]) {
    const outputs = await bundle.generate(outputOptions)
    for (const output of outputs.output as any[]) {
      const { format } = outputOptions
      if (output.code != null) {
        results.push({ format, code: output.code })
      } else {
        results.push({ format, ...output })
      }
    }
  }

  return results
}

afterEach(async () => {
  await fs.remove('lib')
  process.chdir(__dirname)
})

describe('build', function () {
  // Timeout: 60s
  jest.setTimeout(60000)

  test('simple', async function () {
    const caseDir = resolveCaseDir('simple')
    process.chdir(caseDir)
    const results = await build(['fs-extra', '@guanghechen/rollup-config'])
    for (const result of results) {
      const { format, ...output } = result as any
      const data = output.code != null ? output.code : output
      expect(data).toMatchSnapshot(`rollup ${format}`)
    }
    // expect(
    //   fs.existsSync(path.join(caseDir, 'lib/types/index.d.ts')),
    // ).toBeTruthy()
  })
})
