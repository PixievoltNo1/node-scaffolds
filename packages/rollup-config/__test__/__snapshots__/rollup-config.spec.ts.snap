// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build simple: rollup cjs 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var fs = require('fs-extra');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var fs__default = /*#__PURE__*/_interopDefaultLegacy(fs);

var data = {
  name: \\"alice\\"
};
var data$1 = {
  data: data
};

function isFileSync(filepath) {
    if (!fs__default[\\"default\\"].existsSync(filepath))
        return false;
    return fs__default[\\"default\\"].statSync(filepath).isFile();
}

console.log('data:', data$1);
console.log(isFileSync(__filename));
var index = (function () { return data$1; });

exports[\\"default\\"] = index;
exports.isFileSync = isFileSync;
"
`;

exports[`build simple: rollup es 1`] = `
"import fs from 'fs-extra';

var data = {
  name: \\"alice\\"
};
var data$1 = {
  data: data
};

function isFileSync(filepath) {
    if (!fs.existsSync(filepath))
        return false;
    return fs.statSync(filepath).isFile();
}

console.log('data:', data$1);
console.log(isFileSync(__filename));
var index = (function () { return data$1; });

export { index as default, isFileSync };
"
`;

exports[`build simple: rollup es 2`] = `
"/**
 * Check if the given filepath is an existed path of file.
 *
 * @param filepath
 * @returns
 */
declare function isFileSync(filepath: string): boolean;

declare const _default: () => {
    data: {
        name: string;
    };
};

export { _default as default, isFileSync };
"
`;
